libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -389,23 +399,55 @@ void ContractCompiler::appendFunctionSelector(ContractDefinition const& _contrac
 		appendInternalSelector(callDataUnpackerEntryPoints, sortedIDs, notFound, m_optimiserSettings.expectedExecutionsPerDeployment);
 	}
 
-	m_context << notFound;
+	m_context << notFoundOrReceiveEther;
 
-	if (fallback)
+	if (!fallback && !etherReceiver)
+		m_context.appendRevert();
+	else
 	{
-		solAssert(!_contract.isLibrary(), "");
-		if (!fallback->isPayable() && needToAddCallvalueCheck)
-			appendCallValueCheck();
+		std::optional<eth::AssemblyItem> receiveEther;
 
-		solAssert(fallback->isFallback(), "");
-		solAssert(FunctionType(*fallback).parameterTypes().empty(), "");
-		solAssert(FunctionType(*fallback).returnParameterTypes().empty(), "");
-		fallback->accept(*this);
-		m_context << Instruction::STOP;
+		if (etherReceiver)
+		{
+			receiveEther = m_context.newTag();
+			// directly jump to receive ether function, if there is no calldata
+			m_context << Instruction::CALLDATASIZE << Instruction::ISZERO;
+			m_context.appendConditionalJumpTo(*receiveEther);
+		}
+
+		m_context << notFound;
+		if (fallback)
+		{
+			solAssert(!_contract.isLibrary(), "");
+			if (!fallback->isPayable() && needToAddCallvalueCheck)
+				appendCallValueCheck();
+
+			solAssert(fallback->isFallback(), "");
+			solAssert(FunctionType(*fallback).parameterTypes().empty(), "");
+			solAssert(FunctionType(*fallback).returnParameterTypes().empty(), "");
+			fallback->accept(*this);
+			m_context << Instruction::STOP;
+		}
+		else
+			// TODO: error message here?
```

Comment:
Why error message?

Reply:
Inherited form the former version... I think a potential revert message was meant, but not sure.

Reply:
Yep, this is about a revert error message. We can soon add all of them :)

Reply:
Ah ok. Will leave the comment there.